### PR TITLE
GH-113710: Backedge counter improvements.

### PR DESCRIPTION
**Pull request by markshannon**
The PR clarifies the handling of counters and triggering of the optimizer in `JUMP_BACKWARD`.

This PR:
* Clarifies the manual 2s complement conversion.
* Avoids relying on truncation to 16 bits
* Does the shifting and offset calculation when setting the optimizer, rather than on every back edge.
* Hides the handling of the NULL/default optimizer.

* Issue: gh-113710

### CODE REVIEW

File: Python/pystate.c [hunk 1]
```diff
@@ -622,9 +622,7 @@ init_interpreter(PyInterpreterState *interp,
     }
     interp->sys_profile_initialized = false;
     interp->sys_trace_initialized = false;
-    interp->optimizer = &_PyOptimizer_Default;
-    interp->optimizer_backedge_threshold = _PyOptimizer_Default.backedge_threshold;
-    interp->optimizer_resume_threshold = _PyOptimizer_Default.backedge_threshold;
+    _Py_SetOptimizer(interp, NULL);
```
Comment (erlend-aasland) [inline Python/pystate.c]: Hm, how about?

```suggestion
    _PyOptimizerObject *old = _Py_SetOptimizer(interp, NULL);
    assert(old == NULL);
```

...or:

```suggestion
    (void)_Py_SetOptimizer(interp, NULL);
```

Comment (markshannon) [reply Python/pystate.c]: I'll go with your second suggestion. Avoiding compiler warnings for variables only used in asserts is a pain.

**Comment by gvanrossum** [comment]
@markshannon FWIW, even after this PR I find it hard to follow how the Tier 2 optimization backoff counters are managed. The bit fiddling is exceedingly complicated. Maybe we should have some macros for these, like we do for specialization counters.
